pkg/agent/billing/prommetrics.go

```diff
@@ -0,0 +1,99 @@
+package billing
+
+// Prometheus metrics for the agent's billing subsystem
+
+import (
+	"strconv"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	vmapi "github.com/neondatabase/autoscaling/neonvm/apis/neonvm/v1"
+)
+
+type PromMetrics struct {
+	vmsProcessedTotal *prometheus.CounterVec
+	vmsCurrent        *prometheus.GaugeVec
+	batchSizeCurrent  prometheus.Gauge
+	sendErrorsTotal   prometheus.Counter
+}
+
+func NewPromMetrics() PromMetrics {
+	return PromMetrics{
+		vmsProcessedTotal: prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_billing_vms_processed_total",
+				Help: "Total number of times the autoscaler-agent's billing subsystem processes any VM",
+			},
+			[]string{"is_endpoint", "autoscaling_enabled", "phase"},
+		),
+		vmsCurrent: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: "autoscaling_agent_billing_vms_current",
+				Help: "Total current VMs visible to the autoscaler-agent's billing subsystem, labeled by some bits of metadata",
+			},
+			[]string{"is_endpoint", "autoscaling_enabled", "phase"},
+		),
+		batchSizeCurrent: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: "autoscaling_agent_billing_batch_size",
+				Help: "Size of the billing subsystem's most recent batch",
+			},
+		),
+		sendErrorsTotal: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_billing_send_errors_total",
+				Help: "Total errors from attempting to send billing events",
+			},
+		),
+	}
+}
+
+func (m PromMetrics) MustRegister(reg *prometheus.Registry) {
```

Comment:
can you register metrics in the constructor?

Reply:
Technically, yes. TBH though, I'd rather keep it similar to `util/watch` metrics (see also https://github.com/neondatabase/autoscaling/pull/306#discussion_r1214702650)

Reply:
I don't find this pattern/argument particularly compelling for either case, but I don't think this is important enough to reach consensus on.